hpy/tools/autogen/trace.py

```diff
@@ -0,0 +1,200 @@
+        w('}')
+        return '\n'.join(lines)
+
+
+class autogen_tracer_wrappers(AutoGenFile):
+    PATH = 'hpy/trace/src/autogen_trace_wrappers.c'
+
+    def generate(self):
+        lines = []
+        w = lines.append
+        w('#include "trace_internal.h"')
+        w('')
+        for func in self.api.functions:
+            debug_wrapper = self.gen_trace_wrapper(func)
+            if debug_wrapper:
+                w(debug_wrapper)
+                w('')
+        return '\n'.join(lines)
+
+    def gen_trace_wrapper(self, func):
+        if func.name in NO_WRAPPER:
+            return
+
+        assert not func.is_varargs()
+        node = get_trace_wrapper_node(func)
+        const_return = get_return_constant(func)
+        if const_return:
+            make_void(node)
+        signature = toC(node)
+        rettype = get_context_return_type(node, const_return)
+
+        def get_params():
+            lst = []
+            for p in node.type.args.params:
+                if p.name == 'ctx':
+                    lst.append(f'{UCTX_ARG}')
+                else:
+                    lst.append(p.name)
+            return ', '.join(lst)
+        params = get_params()
+
+        lines = []
+        w = lines.append
+        w(signature)
+        w('{')
+        w(f'    HPyTraceInfo *tctx_info = get_info({TCTX_ARG});')
+        w(f'    HPyContext *{UCTX_ARG}= tctx_info->{UCTX_ARG};')
+        w(f'    struct timespec _ts_start, _ts_end;')
+        w(f'    int cr;')
+        w(f'    HPy trace_func_args = HPy_NULL;')
+        w(f'    tctx_info->call_counts[{func.ctx_index}]++;')
+        w(f'    if(!HPy_IsNull(tctx_info->on_enter_func)) {{')
+        w(f'        trace_func_args = hpy_trace_create_func_args({UCTX_ARG}, {func.ctx_index});')
+        w(f'        hpy_trace_on_enter(tctx_info, {UCTX_ARG}, trace_func_args);')
+        w(f'    }}')
+        w(f'    cr = clock_gettime(CLOCK_MONOTONIC_RAW, &_ts_start);')
+        if rettype == 'void':
+            w(f'    {func.name}({params});')
+        else:
+            w(f'    {rettype} res = {func.name}({params});')
+        w(f'    cr += clock_gettime(CLOCK_MONOTONIC_RAW, &_ts_end);')
+        w(f'    if (cr)')
+        w(f'        HPy_FatalError({UCTX_ARG}, "could not get monotonic clock");')
+        w(f'    int64_t duration = diff_ns(_ts_start, _ts_end);')
+        w(f'    assert(duration >= 0);')
+        w(f'    tctx_info->durations[{func.ctx_index}] += duration;')
+        w(f'    if(!HPy_IsNull(tctx_info->on_exit_func)) {{')
+        w(f'        if( HPy_IsNull(trace_func_args))')
+        w(f'            trace_func_args = hpy_trace_create_func_args({UCTX_ARG}, {func.ctx_index});')
+        w(f'        hpy_trace_on_exit(tctx_info, {UCTX_ARG}, trace_func_args);')
+        w(f'    }}')
+        w(f'    HPy_Close({UCTX_ARG}, trace_func_args);')
+        if rettype != 'void':
+            w(f'    return res;')
+        w('}')
+        return '\n'.join(lines)
```

Comment:
I pulled out everything from the wrappers as far as possible except the calls to query the clock. This is because I think there should be no indirection. Please check if you are satisfied.

Reply:
yes, much better now! Thanks